### PR TITLE
Support Quobyte as StorageClass

### PR DESCRIPTION
**Pull request by johscheuer**
This PR allows Users to use Quobyte as StorageClass for dynamic volume provisioning and implements the Provisioner/Deleter Interface. 

@quolix @kubernetes/sig-storage @rootfs

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31434)

### CODE REVIEW

File: pkg/volume/quobyte/quobyte.go [hunk 1]
```diff
@@ -229,19 +251,20 @@ func (mounter *quobyteMounter) SetUpAt(dir string, fsGroup *int64) error {
 // Returns a path in the format ../user@volume e.g. ../root@MyVolume
 // or if a group is set ../user#group@volume
 func (quobyteVolume *quobyte) GetPath() string {
-	user := quobyteVolume.user
-	if len(user) == 0 {
+	var user, group string
+
+	if user = quobyteVolume.user; len(user) == 0 {
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: Prefer:

```go
user := quobyteVolume.user
if len(user) == 0 {
  user = "root"
}
...
```

File: pkg/volume/quobyte/quobyte.go [hunk 2]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+type quobyteVolumeProvisioner struct {
+	*quobyteMounter
+	options volume.VolumeOptions
+}
+
+func (provisioner *quobyteVolumeProvisioner) Provision() (*api.PersistentVolume, error) {
+	if provisioner.options.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+	var err error
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: don't pre-declare things if you don't have to (and I don't think you have to here)

File: pkg/volume/quobyte/quobyte.go [hunk 3]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+		case "user":
+			provisioner.user = v
+		case "group":
+			provisioner.group = v
+		default:
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, provisioner.plugin.GetPluginName())
+		}
+	}
+
+	if user, password, err = provisioner.plugin.getUserAndPasswordFromSecret(adminSecretNamespace, adminSecretName); err != nil {
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: prefer:
```go
user, password, err := ...
if err != nil {
...
}
```

File: pkg/volume/quobyte/quobyte.go [hunk 4]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+		glog.Errorf("failed to get admin secret from [%q/%q]", adminSecretNamespace, adminSecretName)
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: Don't log errors if you are also returning errors (assume someone else will log them at the call site)

File: pkg/volume/quobyte/quobyte.go [hunk 5]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+		return nil, fmt.Errorf("failed to get admin secret from [%q/%q]", adminSecretNamespace, adminSecretName)
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: Don't wrap errors, just return `err`

File: pkg/volume/quobyte/quobyte.go [hunk 6]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+		user:      user,
+		password:  password,
+	}
+	manager := &quobyteVolumeManager{
+		config: cfg,
+	}
+	vol, sizeGB, err := manager.createVolume(provisioner)
+
+	if err != nil {
+		glog.Errorf("quobyte: create volume failed, err: %v", err)
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: Don't log errors and return.

File: pkg/volume/quobyte/quobyte.go [hunk 7]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+		return nil, fmt.Errorf("quobyte: create volume failed, err: %v", err)
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: Don't wrap errors.

File: pkg/volume/quobyte/quobyte.go [hunk 8]
```diff
@@ -258,3 +281,145 @@ func (unmounter *quobyteUnmounter) TearDown() error {
...
+}
+
+func (deleter *quobyteVolumeDeleter) GetPath() string {
+	return deleter.quobyte.GetPath()
+}
+
+func (deleter *quobyteVolumeDeleter) Delete() error {
+	cfg, err := parseVolumeAnnotations(deleter.pv)
+	if err != nil {
+		return nil
```
Comment (brendandburns) [inline pkg/volume/quobyte/quobyte.go]: Shouldn't this be `return err`?

File: vendor/github.com/gorilla/rpc/v2/map.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+		}
+		s.methods[method.Name] = &serviceMethod{
+			method:    method,
+			argsType:  args.Elem(),
+			replyType: reply.Elem(),
+		}
+	}
+	if len(s.methods) == 0 {
+		return fmt.Errorf("rpc: %q has no exported methods of suitable type",
+			s.name)
```
Comment (brendandburns) [inline vendor/github.com/gorilla/rpc/v2/map.go]: move this up to the preceding line.

File: vendor/github.com/gorilla/rpc/v2/server.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
+// Copyright 2009 The Go Authors. All rights reserved.
```
Comment (brendandburns) [inline vendor/github.com/gorilla/rpc/v2/server.go]: Pull this out as a separate commit.

File: vendor/github.com/gorilla/rpc/LICENSE [hunk 11]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2012 Rodrigo Moraes. All rights reserved.
```
Comment (brendandburns) [inline vendor/github.com/gorilla/rpc/LICENSE]: This seems like a bunch of import for a couple of small methods.  Any way we could just write new functions for encode/decode?

Comment (johscheuer) [inline vendor/github.com/gorilla/rpc/LICENSE]: Thanks for the fast review! I updated the code.

If there is a need for others to use similar functions (RPC/JSON calls) we could move this into some util package.

**Comment by rootfs** [comment]
Review status: 0 of 23 files reviewed at latest revision, 16 unresolved discussions.

---

*[examples/experimental/persistent-volume-provisioning/README.md, line 146 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2NJkK7VwoBjs4948r:-KQ2NJkK7VwoBjs4948s:b-7q9v8n) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/README.md#L146)):*
> ```Markdown
> 
> 
> - **apiServer** API Server of Quobyte in the format http(s)://api-server:7860
> ```

apiServer is rather ambiguous, make it quobyteApiServer or something that is distinguishable from kube-apiserver. 

---

*[examples/experimental/persistent-volume-provisioning/README.md, line 147 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2NbskG7OS_lBOm8j7:-KQ2NbskG7OS_lBOm8j8:b-wqs16w) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/README.md#L147)):*
> ```Markdown
> 
> - **apiServer** API Server of Quobyte in the format http(s)://api-server:7860
> * **registry** Quobyte registry to use to mount the volume. You can specifiy the registry as <host>:<port> pair or if you want to specify multiple registries you just have to put a semicolon between them e.q. <host1>:<port>,<host2>:<port>,<host3>:<port>. The host can be an IP address or if you have a working DNS you can also provide the DNS names.
> ```

semicolon->comma

---

*[examples/experimental/persistent-volume-provisioning/README.md, line 150 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2OAYIK2CXnfrsg4GN:-KQ2OAYIK2CXnfrsg4GO:b-wffjg9) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/README.md#L150)):*
> ```Markdown
> * **adminSecretName** secret that holds information about the Quobyte user and the password to authenticate agains the API server.
> * **adminSecretNamespace** The namespace for **adminSecretName**. Default is `default`.
> * **user** maps all access to this user. Default is `root`.
> ```

what does "all access" mean?

---

*[examples/experimental/persistent-volume-provisioning/quobyte/quobyte-storage-class.yaml, line 7 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2NxNe8Kl-1Kso1YLi:-KQ2NxNe8Kl-1Kso1YLj:b-wt8d1i) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/quobyte/quobyte-storage-class.yaml#L7)):*
> ```YAML
> provisioner: kubernetes.io/quobyte
> parameters:
>     apiServer: http://139.59.210.140:7860
> ```

https if possible

---

*[pkg/volume/quobyte/quobyte.go, line 50 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2OZ4ndjDqkuEOK5n-:-KQ2OZ4ndjDqkuEOK5n0:ba7g3x1) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L50)):*
> ```Go
> 	user      string
> 	password  string
> 	apiServer string
> ```

quobyteApiServer for less confusion, same for the rest.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434)*

File: vendor/github.com/gorilla/rpc/LICENSE (hunk 11 above)
Comment (brendandburns) [inline vendor/github.com/gorilla/rpc/LICENSE]: My point is that maybe we should just implement our own encode/decode rather than importing ~500 lines of code we're barely going to use...

Comment (brendandburns) [inline vendor/github.com/gorilla/rpc/LICENSE]: fwiw, I don't super care, but I'm trying to limit dependency sprawl from encompassing the whole internet.

**Comment by johscheuer** [comment]
Review status: 0 of 23 files reviewed at latest revision, 16 unresolved discussions.

---

*[examples/experimental/persistent-volume-provisioning/README.md, line 146 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2NJkK7VwoBjs4948r:-KQ5_XZZieoN1d3d6QCj:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/README.md#L146)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary>
> apiServer is rather ambiguous, make it quobyteApiServer or something that is distinguishable from kube-apiserver. 
</details>
Done.

---

*[examples/experimental/persistent-volume-provisioning/README.md, line 147 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2NbskG7OS_lBOm8j7:-KQ5_YyV10-2W1tD0l9_:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/README.md#L147)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary>
> semicolon->comma
</details>
Done.

---

*[examples/experimental/persistent-volume-provisioning/README.md, line 150 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2OAYIK2CXnfrsg4GN:-KQ5a15l22pttgmaDihV:b-5dblfx) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/README.md#L150)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary>
> what does "all access" mean?
</details>
All (read/write) access from the Container(s) to the files inside the volume independent from the user inside the Container.

---

*[examples/experimental/persistent-volume-provisioning/quobyte/quobyte-storage-class.yaml, line 7 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2NxNe8Kl-1Kso1YLi:-KQ5_cznqgiDSUAWxdIL:btlqe0z) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/examples/experimental/persistent-volume-provisioning/quobyte/quobyte-storage-class.yaml#L7)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary>
> https if possible
</details>
Done. It's still only an example you can talk with the API Server with HTTP or HTTPS (expects a valid certificate)

---

*[pkg/volume/quobyte/quobyte.go, line 50 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ2OZ4ndjDqkuEOK5n-:-KQ5aOkQfutoCMWPUuN3:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L50)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary>
> quobyteApiServer for less confusion, same for the rest.
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 256 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-256:-KQ5aL-k6JSkRterHL_9:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L256)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Prefer:
> 
> 
> 
> ```go
> 
> user := quobyteVolume.user
> 
> if len(user) == 0 {
> 
>   user = "root"
> 
> }
> 
> ...
> 
> ```
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 348 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-348:-KQ5aL_ds3QdYr8_Z5D6:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L348)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> don't pre-declare things if you don't have to (and I don't think you have to here)
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 371 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-371:-KQ5aM596Dqsz0zAQq2D:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L371)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> prefer:
> 
> ```go
> 
> user, password, err := ...
> 
> if err != nil {
> 
> ...
> 
> }
> 
> ```
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 372 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-372:-KQ5abbTRsSRidgvBfEx:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L372)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Don't log errors if you are also returning errors (assume someone else will log them at the call site)
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 373 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-373:-KQ5aN4UPuNyBAzNjVf1:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L373)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Don't wrap errors, just return `err`
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 398 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-398:-KQ5aNctg9GUWB5KXFwD:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L398)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Don't log errors and return.
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 399 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-399:-KQ5aNzltaetYACOsYHN:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L399)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Don't wrap errors.
</details>
Done.

---

*[pkg/volume/quobyte/quobyte.go, line 420 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoAK6GZrHonRazH-r1-420:-KQ5aOJ5EgwkHv3_nZAl:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/pkg/volume/quobyte/quobyte.go#L420)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Shouldn't this be `return err`?
</details>
Done.

---

*[vendor/github.com/gorilla/rpc/LICENSE, line 1 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoBAGaT2h_rz8Pn-r1-1:-KQ5aRcOUUyWShyCyRys:b3u1kv4) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/vendor/github.com/gorilla/rpc/LICENSE#L1)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> fwiw, I don't super care, but I'm trying to limit dependency sprawl from encompassing the whole internet.
</details>
Sounds reasonable, I will have a look at it

---

*[vendor/github.com/gorilla/rpc/v2/server.go, line 1 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434#-KQ1DDoN7_fEFfuRER_V-r1-1:-KQ5a_LKqPhEk49RPFgg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ac36ac1004f803e83500ff1daf62c9f97e7a2e3a/vendor/github.com/gorilla/rpc/v2/server.go#L1)):*
<details><summary><i>Previously, brendandburns (Brendan Burns) wrote…</i></summary>
> Pull vendor code out as a separate commit.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31434)*

**Comment by johscheuer** [comment]
@brendandburns we update the Quobyte API client and remove the external `github.com/gorilla` dependencies from it.

@rootfs got some time for reviewing the code?

**Comment by brendandburns** [comment]
I still see the gorilla library being added.  Missing a push?  or perhaps just need to remove the commit?

Thanks

**Comment by johscheuer** [comment]
Sorry I didn't make it earlier, now the code base is updated and the gorilla dependency is removed.

**Comment by brendandburns** [comment]
This LGTM, but I defer to @rootfs since he also had comments.

**Comment by rootfs** [comment]
@johscheuer @brendandburns Code looks fine but it is yet to be decided the best way to deal with secret. See discussions at https://github.com/kubernetes/kubernetes/pull/31869

**Comment by johscheuer** [comment]
@rootfs I would adjust the PR to save only the secret in the Volume annotations (I like this a way more than saving the plain password) and then obviously we have to wait till there is a decision for the secrets.

**Comment by rootfs** [comment]
@johscheuer will keep you updated.

**Comment by johscheuer** [comment]
@rootfs I added quoybte Config and Tenant (nothing special) https://github.com/kubernetes/kubernetes/pull/31434/files#diff-18465b1c330b6ccd5a88505ad05cca5bR167

Do you know if we discuss the secret issue at the sig-storage meeting?

**Comment by johscheuer** [comment]
@rootfs ah cool! I didn't saw that you added the GetSecret method -> https://github.com/kubernetes/kubernetes/pull/31251/files#diff-c3acf7a7fec401d0a3cb4998088ee1e5R114 I will adjust my PR.

File: pkg/volume/quobyte/quobyte_util.go [hunk 12]
```diff
@@ -46,3 +93,53 @@ func (mounter *quobyteMounter) pluginDirIsMounted(pluginDir string) (bool, error
...
+		if _, _, err := net.SplitHostPort(hostPortPair); err != nil {
+			return false
+		}
+	}
+
+	return true
+}
+
+// ADD Annotations to a PersistentVolume all information needed to delete the Volume
+func addVolumeAnnotations(apiServer, adminSecret, adminSecretNamespace string, pv *api.PersistentVolume) {
```
Comment (johscheuer) [inline pkg/volume/quobyte/quobyte_util.go]: @rootfs @jsafrane what about adding a generic method `func addVolumeAnnotations(params map[string]string, pv *api.PersistentVolume)` to the `volume/util`, similar to https://github.com/kubernetes/kubernetes/pull/31251/files#diff-c3acf7a7fec401d0a3cb4998088ee1e5R114 ? The same for `parseVolumeAnnotations` (`func parseVolumeAnnotations(pv *api.PersistentVolume) (map[string]string, error)`)? This would remove the need that every Volume Plugin writes it own methods for Annotations adding/parsing. 

If you think this would be an improvement I would add this change in this PR (or an separated if you want).

Comment (rootfs) [inline pkg/volume/quobyte/quobyte_util.go]: Yes we do have a common case of using pv annotation. Please refactor as needed. Thanks

Comment (johscheuer) [inline pkg/volume/quobyte/quobyte_util.go]: Updated the code

File: pkg/volume/util/util.go [hunk 13]
```diff
@@ -126,3 +127,33 @@ func GetSecret(namespace, secretName string, kubeClient clientset.Interface) (ma
...
 }
+
+// AddVolumeAnnotations adds a golang Map as annotation to a PersistentVolume
+func AddVolumeAnnotations(pv *api.PersistentVolume, annotations map[string]string) {
+	if pv.Annotations == nil {
+		pv.Annotations = map[string]string{}
+	}
+
+	for k, v := range annotations {
+		pv.Annotations[k] = v
```
Comment (johscheuer) [inline pkg/volume/util/util.go]: Is there a need to check if theres already a value set for the specified key?

Comment (rootfs) [inline pkg/volume/util/util.go]: the annotation should be empty when the pv is provisioned.

**Comment by johscheuer** [comment]
@rootfs can we tag this PR to be tested?

So in the first step this PR is fine with a reference to the secrets in the PV Annotations but in a second step (when external provisioners  are available) we should refactor this into an external provisioners?

**Comment by jsafrane** [comment]
@k8s-bot ok to test

**Comment by jsafrane** [comment]
looks OK to me, @brendandburns, @rootfs opinions?

@johscheuer, can you please provide an e2e tests for Quobyte volume plugin (mount/unmount) in a subsequent PR? Look at `test/e2e/volumes.go` for an example how to test with containerized storage clusters like Gluster or Ceph - typically some minimal (and often also degraded) server is started as a pod, all we need is that it provides a minimal volume that mounts.

Test for provisioning would be helpful too (`test/e2e/volume_provisioning.go`).

**Comment by johscheuer** [comment]
@jsafrane I will have a look at but only in two weeks I'm stuffed with other things at the moment that I have to finish first. When I correctly remember we decided to not implement a 2e2 for Quobyte these reasons: https://github.com/kubernetes/kubernetes/pull/24977#issuecomment-219657754 (+following)

**Comment by rootfs** [comment]
LGTM, pass @brendandburns for tags.

**Comment by brendandburns** [comment]
tagged.

**Comment by johscheuer** [comment]
@brendandburns looks like my last push got stuck :( I fixed a small typo before the `lgtm` and I thought my computer had pushed it looks like I was wrong.. Sorry!

**Comment by jsafrane** [comment]
@k8s-bot test this issue #32770

**Comment by johscheuer** [comment]
Thanks!
